### PR TITLE
Validator Re-Enabling (master PR)

### PR DESCRIPTION
**Pull request by Overkillus**
Aims to implement Stage 3 of Validator Disbling as outlined here: https://github.com/paritytech/polkadot-sdk/issues/4359

### CODE REVIEW

File: substrate/frame/staking/src/lib.rs [hunk 1]
```diff
@@ -1328,18 +1355,97 @@ impl<T: Config, const DISABLING_LIMIT_FACTOR: usize> DisablingStrategy<T>
...
+			log!(
+				debug,
+				"Reached disabling limit {:?}, checking for re-enabling",
+				Self::disable_limit(active_set.len())
+			);
+
+			// Find the smallest offender to re-enable that is not higher than offender_slash_severity
+			if let Some((smallest_idx, _)) = currently_disabled
+				.iter()
+				.filter(|(_, perbill)| *perbill <= offender_slash_severity)
```
Comment (tdimitrov) [inline substrate/frame/staking/src/lib.rs:1436]: Just to clarify - even if there is no offender with a smaller offence we will re-enable the first guy with an offence equal to the new offender. Which is a good thing.

However I think we are biased here towards disabling the validator with the smallest index, which is wrong. If there are multiple elements with the same min value `min_by_key` returns the first one. If I am not mistaken `DisabledValidators` are sorted by key ([here](https://github.com/paritytech/polkadot-sdk/pull/5724/files#diff-8dd07463cc135387ca715a01d60b347276b1ff0b5e73db9dc4e8a1390fb462ceR2303)) so this way we will always reenable the guy with the smallest index.

It's a corner case but it would be nice to handle it if it is not too complicated.

File: substrate/frame/staking/src/slashing.rs [hunk 2]
```diff
@@ -321,23 +321,52 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
...
-			}
-		}
-	});
-
-	// `DisabledValidators` should be kept sorted
-	debug_assert!(DisabledValidators::<T>::get().windows(2).all(|pair| pair[0] < pair[1]));
+    DisabledValidators::<T>::mutate(|disabled| {
+        let decision = T::DisablingStrategy::decision(params.stash, params.slash, params.slash_era, &disabled);
+
+        match (decision.disable, decision.reenable) {
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:329]: Probably I am missing something but won't this be simpler if the two values are handled separately?
I mean if `decision.disable` is `Some` - disable the guy. Then if `decision.reenable` is `Some` - reenable the guy?

That way you will avoid the code duplication.

File: substrate/frame/staking/src/slashing.rs [hunk 3]
```diff
@@ -321,23 +321,52 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
...
+					T::SessionInterface::enable_validator(reenable_idx);
+                }
+
+                // Add the validator to `DisabledValidators` and disable it. Do nothing if it is
+                // already disabled.
+                if let Err(index) = disabled.binary_search_by_key(&offender_idx, |(index, _)| *index) {
+					let severity = params.slash;
+                    disabled.insert(index, (offender_idx, severity));
+                    // Propagate disablement to session level
+                    T::SessionInterface::disable_validator(offender_idx);
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:358]:
```suggestion
                    T::SessionInterface::disable_validator(offender_idx);
                    
                    todo!("Add severity updating")
```

**Review by tdimitrov — COMMENTED** [review]
I had a quick pass by focusing mainly on the approach. It looks good, nice work @Overkillus!

I've left some thoughts about a corner case with the re-enabling.

File: substrate/frame/staking/src/slashing.rs (hunk 2 above)
Comment (Overkillus) [reply substrate/frame/staking/src/slashing.rs]: Initially opted for this approach as this was alliged with all the possible outputs of decision. Just for readability and explicit separate handling. Maybe it was an overkill so switched back to your suggestion.

File: substrate/frame/staking/src/lib.rs (hunk 1 above)
Comment (Overkillus) [reply substrate/frame/staking/src/lib.rs:1454]: You are definitely right. I simply don't think this is anything to worry about. This will only kick in and matter only when we have offender of the same level of severity. Slashes will be dispensed and the newest offender will be disabled. From our perspective as long as lowest severity is re-enabled this is the lowest possible risk. 

We will mention it to SRLabs on audit and see what the think. We can remedy as you mentioned with a random selection for a given severity level instead of arbitrary.

File: substrate/frame/staking/src/slashing.rs [hunk 4]
```diff
@@ -321,17 +321,48 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
...
+					});
+				},
+			}
+		}
+
+		if let Some(reenable_idx) = decision.reenable {
+			// Remove the validator from `DisabledValidators` and re-enable it.
+			if let Ok(index) = disabled.binary_search_by_key(&reenable_idx, |(index, _)| *index) {
+				disabled.remove(index);
+				// Propagate re-enablement to session level
```
Comment (ordian) [inline substrate/frame/staking/src/slashing.rs:358]: i got confused by having the same storage name `DisabledValidators` in slashing and in session. maybe we could rename the one in staking to `DisabledValidatorsWithSeverity` or something like that

**Review by ordian — COMMENTED** [review]
Looks good overall, but spotted some things that need to be fixed

File: substrate/frame/staking/src/lib.rs [hunk 5]
```diff
@@ -1328,18 +1350,117 @@ impl<T: Config, const DISABLING_LIMIT_FACTOR: usize> DisablingStrategy<T>
...
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached re-enabling logic might kick in.
+	pub fn disable_limit(validators_len: usize) -> usize {
+		validators_len
+			.saturating_sub(1)
+			.checked_div(DISABLING_LIMIT_FACTOR)
+			.unwrap_or_else(|| {
+				defensive!("DISABLING_LIMIT_FACTOR should not be 0");
+				0
+			})
```
Comment (tdimitrov) [inline substrate/frame/staking/src/lib.rs:1388]: nit: Since this logic is shared between `UpToLimitWithReEnablingDisablingStrategy` and `UpToLimitDisablingStrategy` it's worth to extract it in a separate function.

File: substrate/frame/staking/src/tests.rs [hunk 6]
```diff
@@ -8277,48 +8301,433 @@ mod byzantine_threshold_disabling_strategy {
...
+			assert!(pre_upgrade_result.is_ok());
+			let pre_upgrade_state = pre_upgrade_result.unwrap();
+
+			// Run the migration
+			v16::VersionUncheckedMigrateV15ToV16::<Test>::on_runtime_upgrade();
+
+			// Run post-upgrade checks
+			let post_upgrade_result =
+				v16::VersionUncheckedMigrateV15ToV16::<Test>::post_upgrade(pre_upgrade_state);
+			assert!(post_upgrade_result.is_ok());
```
Comment (tdimitrov) [inline substrate/frame/staking/src/tests.rs:8730]: This is not needed. The same steps are performed by `try-runtime` in a CI job.

File: substrate/frame/staking/src/pallet/mod.rs [hunk 7]
```diff
@@ -727,7 +727,7 @@ pub mod pallet {
 	/// offended using binary search.
 	#[pallet::storage]
 	#[pallet::unbounded]
-	pub type DisabledValidators<T: Config> = StorageValue<_, Vec<u32>, ValueQuery>;
+	pub type DisabledValidators<T: Config> = StorageValue<_, Vec<(u32, Perbill)>, ValueQuery>;
```
[the remark this answers is not in the document]
Comment (Overkillus) [reply substrate/frame/staking/src/pallet/mod.rs]: Switched to an OffenceSeverity wrapper with default ord impl for now but can be readjusted if need be.

File: substrate/frame/staking/src/pallet/mod.rs [hunk 8]
```diff
@@ -723,11 +724,14 @@ pub mod pallet {
 	/// implementor of [`DisablingStrategy`] defines if a validator should be disabled which
 	/// implicitly means that the implementor also controls the max number of disabled validators.
 	///
-	/// The vec is always kept sorted so that we can find whether a given validator has previously
-	/// offended using binary search.
+	/// The vec is always kept sorted based on the u32 ubdex so that we can find whether a given
```
Comment (ordian) [inline substrate/frame/staking/src/pallet/mod.rs:727]:
```suggestion
	/// The vec is always kept sorted based on the u32 index so that we can find whether a given
```

File: polkadot/runtime/westend/src/lib.rs [hunk 9]
```diff
@@ -162,7 +162,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("westend"),
 	impl_name: create_runtime_str!("parity-westend"),
 	authoring_version: 2,
-	spec_version: 1_015_000,
+	spec_version: 1_016_002,
```
Comment (ordian) [inline polkadot/runtime/westend/src/lib.rs:165]: this change seems unrelated

File: substrate/frame/staking/src/pallet/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Overkillus) [reply substrate/frame/staking/src/pallet/mod.rs]: Context:
DisabledValidators is stored on two different levels: staking and session. On staking level we now store it with severities and this is where the migration happened. On session level there have been no storage changes. 

So with regards to #64 polkadot runtime parachains only refers to the session disabling validators as the runtime API which AFAIK should be unbroken by the migration. I don't think we expose the staking level disabled validators through runtime apis.

File: polkadot/runtime/westend/src/lib.rs (hunk 9 above)
Comment (sandreim) [reply polkadot/runtime/westend/src/lib.rs:165]: an artifact of testing the runtime upgrade

File: substrate/frame/staking/src/pallet/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ordian) [reply substrate/frame/staking/src/pallet/mod.rs]: Makes sense, thanks! Got confused by same naming: https://github.com/paritytech/polkadot-sdk/pull/5724#discussion_r1800655121

File: substrate/frame/staking/src/tests.rs (hunk 6 above)
Comment (Overkillus) [reply substrate/frame/staking/src/tests.rs:8729]: Are they? I am manually adding some pre-state so that the migration does not run on empty DIsabledValidators list. 

```rust
let old_disabled_validators = vec![1u32, 2u32];
			v15::DisabledValidators::<Test>::put(old_disabled_validators.clone());
```

The basic try-runtime CI job AFAIK only uses pre_upgrade to prepare the state in which is on purpose did not modify the state so it might run the migraiton on empty disabled lists. This test covers the case where the lists are not empty so we can see the migration happening accordingly.

File: substrate/frame/staking/src/slashing.rs [hunk 10]
```diff
@@ -321,17 +321,48 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
...
+		let decision =
+			T::DisablingStrategy::decision(params.stash, params.slash, params.slash_era, &disabled);
+
+		if let Some(offender_idx) = decision.disable {
+			// Check if the offender is already disabled
+			match disabled.binary_search_by_key(&offender_idx, |(index, _)| *index) {
+				// Offender is already disabled, update severity if the new one is higher
+				Ok(index) => {
+					let (_, old_severity) = &mut disabled[index];
+					if params.slash > *old_severity {
```
[the remark this answers is not in the document]
Comment (Overkillus) [reply substrate/frame/staking/src/slashing.rs]: We are indeed only depositing slashes that are the highest in an era. That is a good catch. The [`UpToLimitWithReEnablingDisablingStrategy`] simply considers the most recent given to us slash if it was larger then previous ones seen. So re-enabling has a baked in max assumption.

This means that it would still work on cumulative slashes but possibly not as one intends (AKA only use the highest slash instead of total slashes).

I added this assumption explicitly mentioned in the [`UpToLimitWithReEnablingDisablingStrategy`]

```rust
/// This strategy is not based on cumulative severity of offences but only on the severity of the highest offence.
/// Offender first committing a 25% offence and then a 50% offence will be treated the same as an offender committing 50%
/// offence.
```

If someone would like to have a cumulative disabling version this would require a new disablingStrategy impl.

**Review by ordian — APPROVED** [review]

**Review by tdimitrov — APPROVED** [review]

**Comment by Overkillus** [comment]
bot fmt

**Comment by Ank4n** [comment]
Is it possible to move most of the disabling logic to `pallet_session`? 

It seems staking does not really need to know anything about disabling. Currently, we are maintaing two copies of disabled validator indices in both `pallet_session` and `pallet_staking`, apparently because staking only knows about era, but I am not sure what does that mean. Staking does not have any concept of `disable` in it, and it should only care about slashing. If the concern is that we want to keep a validator disabled for the whole era, staking returns a set of validators only for sessions that trigger an era. For other sessions, it simply returns `None`. So session should know when to clear disabled validators.

There is a larger reason as to why I am flagging this. With staking moving to AH, the offence lifecycle would be something like this:
- Offense is reported and verified on RC.
- Any disable logic applied and propagated by `pallet_session`. It should also keep track of `SlashPerbill` in its `DisabledValidators` store.
- An offence report (Offender, Session, SlashPerbill) is sent to AH for slashing. 

In practice, `pallet_staking` will be replaced by `pallet_staking_client` (or a more suitable name) that will fill in the gaps. That is, it will act as `OnOffenceHandler` and `SessionManager` on RC, and async communicate with Staking/AH.

File: prdoc/pr_5724.prdoc [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+      This PR changes when an active validator node gets disabled for comitting offences.
+      When Byzantine Threshold Validators (1/3) are already disabled instead of no longer
+      disabling the highest offenders will be disabled potentially re-enabling low offenders.
+
+  - audience: Node Operator
+    description: |
+      Implementation of the Stage 3 for the New Disabling Strategy: https://github.com/paritytech/polkadot-sdk/issues/4359
+
+      This PR changes when an active validator node gets disabled within parachain consensus (reduced responsibilities and
```
[the remark this answers is not in the document]
Comment (Overkillus) [reply prdoc/pr_5724.prdoc:19]: Not really. Parachains don't have their own native consensus or security and they derive it from the RLC.

In Polkadot when we say parachain consensus we are referring to parts of polkadot protocol that allow the inheritance of security and consensus to parachains AKA the part where we assign responsibility (backing), ensure availability, recheck correctness (approval checking rerunning of PVs) and punish wrongdoers (disputes and slashes). 

I can maybe change it Parachain Consensus Protocol to make it a bit clearer but otherwise it is consistent with all our other materials. See here for instance: https://wiki.polkadot.network/docs/learn-parachains-faq#:~:text=%22Parachain%20consensus%22%20is%20special%20in,can%20control%20their%20own%20consensus.

**Comment by Overkillus** [comment]
> Is it possible to move most of the disabling logic to pallet_session?

(assuming we want to do this)

I will discuss if this is a good idea later below but even if we assume we want to do this I don't believe it is appropriate to do this in this PR. This PR only aims to, with minimal changes and refactors, simply allow for validator re-enabling using all the previous already in-place logic. I want to make sure that this PR is minimal and adheres to what was already pre-agreed in the design document for disabling, to limit audit time and just be more sure that everything does exactly what we think it does.

If we have plans to refactor the staking pallet (which as you mentioned will be done anyway during the port to AH) then it should be done in a separate primarily refactor/port PR. 

---

> Is it possible to move most of the disabling logic to pallet_session?

(answering if we want to do this)

Staking pallet as it stands is responsible for more than just slashing. It holds important params with regards to the active validator set. For instance min, ideal, and max validator counts. What it means to be disabled is not within the purview of the staking pallet, what it does is simply keeps track of highest offenders up to some limit and makes this info available for others. This is highly customisable and different users of this pallet can add new strategies with new limits.

Since we keep those validators for an era it makes sense to keep track of their offences (potential disabling status) also for a full era. Storing this information in session when we aim to keep it for era seems like an unintuitive approach.

I might agree that we could make it less opinionated by moving disabling strategies to session and instead keeping a history of all offences (unfiltered by disabling strategies) in staking. Nevertheless in this approach we still should keep track of all offences in staking. An offence is more than a slash, there might by other byproducts or consequences and we need to allow for them.

In the new pipeline for offences that you suggested, where do you think a history of offences within an era should be stored?

Session - Staking coupling

It is okay to keep staking bounded to session, but session should not be dependant on staking. Adding logic to session that is actually era-wide in scope through hacky reads of signals from staking seems well, hacky. This information should be kept in a context that is explicitly era-wide and shared with session.

TLDR: This might be a needed change but is outside of the scope of this PR and some of that information should still live in era-wide and not session-wide scopes.

**Comment by Overkillus** [comment]
@Ank4n 

> I wouldn't try to block you if you want to go ahead with this PR, especially since most of the things I flagged should ideally had been flagged with the earlier PRs, and this PR does not introduce a new design decision.

That is what I would suggest. This impl is simply a part of an earlier deisgn which was pre-approved by SRLabs in the design's audit. It just adds some new functionality using the same infrastructure without altering it too much.

> Since this PR touches the logic that we know we will have to migrate in next couple of months, it is reasonable enough enough to make those changes in this PR.

We have the power to open multiple PRs to separate unrelated changes between the PRs. Enabling new functionality in the current design should not come with a major refactor if it is not neccesary. I am open to the refactor at a later stage but nevertheless it should be a separate PR.

This PR does **not** make the future refactor harder or easier. I would understand withholding the change if it made the refactor harder. It does not make it harder and is orthogonal. 

The rest of the discussion dives into why the refactor might be a good idea later on which we should separate into a separate issue or ticket. I'd be happy to participate in those discussions and help as much as I can.

**Comment by tdimitrov** [comment]
> We have the power to open multiple PRs to separate unrelated changes between the PRs. Enabling new functionality in the current design should not come with a major refactor if it is not neccesary. I am open to the refactor at a later stage but nevertheless it should be a separate PR.

I agree with @Overkillus here. Considering that this PR is part of the disabling strategy roll out I am strongly against doing any out of scope re-factorings. 

What @Ank4n suggests definitely makes sense and we should do it but as a separate effort. Let's focus on the disabling strategy in this PR.

**Comment by Ank4n** [comment]
@tdimitrov @sandreim The refactoring needed is not unrelated. But blocking also serves no purpose, and we can handle this in a followup issue. 

@Overkillus Could you look into if its possible to get rid of the `DisabledValidator` in pallet-staking and only maintain it in pallet-session (even if it results in a bit messy code)? This would make the followup refactor noop. Otherwise, I'm ready to approve.

@gpestana You may want to take a quick look at the changes as well.

**Comment by Overkillus** [comment]
Added all the defensive suggestions from @Ank4n Good eye for spotting them, thanks!

> Could you check if its possible to get rid of the storage item DisabledValidators in pallet-staking and only maintain it in pallet-session (even if it results in a bit messy code)?

This is not trivial. 99% of the disabling logic lives in staking so moving it over is not easy and this is a major part of the needed refactor.

**Review by Ank4n — APPROVED** [review]

File: substrate/frame/staking/src/slashing.rs (hunk 4 above)
Comment (Overkillus) [reply substrate/frame/staking/src/slashing.rs:358]: Leaving for now but this will be potentially simplified in a future refactor connected to moving staking to a parachain. It might require moving the disabling logic off the staking pallet and will remove the redundancy.

**Comment by Overkillus** [comment]
Tests in CI failed on [test-linux-stable (1/3, parity-large-persistent)](https://github.com/paritytech/polkadot-sdk/actions/runs/11896057400/job/33147184815#logs):

```
   FLAKY 4/6 [  31.550s] pallet-revive-eth-rpc tests::deploy_and_call
   FLAKY 2/6 [   0.107s] polkadot-dispute-distribution tests::send_dispute_gets_cleaned_up
   FLAKY 2/6 [ 126.884s] polkadot-node-core-pvf::it execute_job_terminates_on_timeout
  TRY 6 FAIL [  30.260s] pallet-revive-eth-rpc tests::native_evm_ratio_works
  TRY 6 FAIL [   2.420s] sc-rpc-spec-v2 chain_head::tests::ensure_operation_limits_works
  ```

- eth-rpc tests (deploy_and_call & native_evm_ratio_works) seem to be very flaky overall, native_evm_ratio_works always fails locally (0/30)
- send_dispute_gets_cleaned_up is a bit weirder, passing locally with no fail (100/100)
- execute_job_terminates_on_timeout, always passes locally (100/100)
- ensure_operation_limits_works, always passes locally (100/100)

Looking into flaking tests further.
